### PR TITLE
Jaliya/chunk distributed

### PR DESCRIPTION
**Pull request by jaliyae**
Adding distributed data loading support to chunk data set. This removes the LockedSampler we needed and added a simple concept called ChunkSelector as part of the chunk dataset.
It also introduced the API set_epoch so that the ChunkSelector can perform shuffling between epochs if needed.

### CODE REVIEW

File: test/cpp/api/dataloader.cpp [hunk 1]
```diff
@@ -1445,17 +1611,19 @@ TEST(DataLoaderTest, ChunkDataSetGetBatch) {
...
-            samplers::SequentialSampler,
             samplers::SequentialSampler>>
             dataset = datasets::make_shared_dataset<datasets::ChunkDataset<
                 DummyChunkDataReader,
-                samplers::SequentialSampler,
                 samplers::SequentialSampler>>(
                 data_reader,
                 sampler,
-                sampler,
+                std::move(chunk_selector),
```
Comment (jaliyae) [inline test/cpp/api/dataloader.cpp]: We don't need to move.

File: test/cpp/api/dataloader.cpp [hunk 2]
```diff
@@ -97,17 +97,19 @@ TEST(DataTest, ChunkDataSetWithInvalidInitParameter) {
 
   auto initialization_function =
       [&](size_t preloader_count, size_t batch_size, size_t cache_size) {
+        std::shared_ptr<datasets::detail::ChunkSelector> chunk_selector =
```
Comment (xzhu1900) [inline test/cpp/api/dataloader.cpp]: If ChunkSelector is going to be a public class, should we move it out of detail namespace?

Comment (xzhu1900) [inline test/cpp/api/dataloader.cpp]: If ChunkSelector is going to be used by user directly (not an internal-use-only class), does it make more sense to move it out of detail namespace?

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 3]
```diff
@@ -240,6 +249,133 @@ class BatchDataBuffer {
...
+    {
+      for (size_t i = chunk_count; i < index_count; ++i) {
+        all_indices_[i] =
+            i % chunk_count_; // we are adding some more chunks to make all
+                              // replicas to have the same number of chunks.
+      }    
+    }
+  }
+
+  optional<size_t> next() {
```
Comment (xzhu1900) [inline torch/csrc/api/include/torch/data/datasets/chunk.h]: Nit: missed an 'override' here.

**Review by jaliyae — COMMENTED** [review]
I agree, in my original version it did not have to, but I think this make sense. I will move it outside.

File: test/cpp/api/dataloader.cpp (hunk 2 above)
Comment (jaliyae) [reply test/cpp/api/dataloader.cpp]: I agree, in my original version it did not have to, but I think this make sense. I will move it outside.

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 4]
```diff
@@ -240,6 +249,133 @@ class BatchDataBuffer {
...
+/// during chunk loading.
+class ChunkSelector {
+ public:
+  virtual ~ChunkSelector() = default;
+  ChunkSelector(size_t chunk_count, size_t num_replicas = 1, size_t rank = 0)
+      : chunk_count_(chunk_count),
+        num_replicas_(num_replicas),
+        rank_(rank),
+        epoch_(0) {
+    local_chunk_count_ = (size_t)std::ceil(chunk_count_ * 1.0 / num_replicas_);
```
Comment (xzhu1900) [inline torch/csrc/api/include/torch/data/datasets/chunk.h]: Nit: prefer static_cast over c style cast.

Comment (jaliyae) [reply torch/csrc/api/include/torch/data/datasets/chunk.h]: fixed.

File: torch/csrc/api/include/torch/data/datasets/chunk.h (hunk 3 above)
Comment (jaliyae) [reply torch/csrc/api/include/torch/data/datasets/chunk.h]: fixed.

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 5]
```diff
@@ -415,14 +552,15 @@ class ChunkDataset final
...
   ChunkReader chunk_reader_;
 
-  // chunk sampler to shuffle different chunks
-  samplers::LockedSampler<ChunkSamplerType> chunk_sampler_;
-
   // example sampler to shuffle examples in a specific chunk
   ExampleSamplerType example_sampler_;
 
+  // Selects chunks and their order for this reader.
+  std::shared_ptr<detail::ChunkSelector> chunk_selector_;
```
Comment (xzhu1900) [inline torch/csrc/api/include/torch/data/datasets/chunk.h]: It looks like ChunkDataset uniquely own this selector. Shall we make it a unique_ptr?

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 6]
```diff
@@ -287,31 +423,28 @@ struct ChunkDatasetOptions {
...
   ChunkDataset(
       ChunkReader chunk_reader,
-      ChunkSampler chunk_sampler,
       ExampleSampler example_sampler,
+      std::shared_ptr<detail::ChunkSelector> chunk_selector,
       ChunkDatasetOptions options)
       : chunk_reader_(std::move(chunk_reader)),
-        chunk_sampler_(std::move(chunk_sampler)),
         example_sampler_(std::move(example_sampler)),
+        chunk_selector_(std::move(chunk_selector)),
```
Comment (xzhu1900) [inline torch/csrc/api/include/torch/data/datasets/chunk.h:457]: Let's add a check here to make sure chunk_selector is not nullptr.

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 7]
```diff
@@ -240,6 +249,133 @@ class BatchDataBuffer {
...
+    end_index_ = begin_index_ + local_chunk_count_;
+    chunk_index_ = begin_index_;
+  }
+
+  optional<size_t> next() {
+    size_t idx = chunk_index_.fetch_add(1, std::memory_order_relaxed);
+    if (idx < end_index_) {
+      return idx % chunk_count_;
+    } else {
+      return nullopt;/// Select chunks randomly. The chunk order shuffled at each `reset()` call
```
Comment (xzhu1900) [inline torch/csrc/api/include/torch/data/datasets/chunk.h]: It seems this comment is stale.

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 8]
```diff
@@ -240,6 +249,133 @@ class BatchDataBuffer {
...
+  }
+
+  void reset() override {
+    chunk_index_ = begin_index_;
+  }
+
+  private:
+  size_t begin_index_;
+  size_t end_index_;
+  std::atomic<size_t> chunk_index_;
```
Comment (xzhu1900) [inline torch/csrc/api/include/torch/data/datasets/chunk.h:390]: I don't know whether it's worth to have another non-atomic version for non-distributed system case. Currently, in both scenarios, every 'next()' call goes through the atomic chunk_index_ variable. The concurrent restriction is not necessary in the non-distributed system and would have a perf hit on the atomic lock.

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 9]
```diff
@@ -240,6 +249,133 @@ class BatchDataBuffer {
...
+    chunk_indices_.clear();
+    chunk_indices_.reserve(local_chunk_count_);
+    auto begin_index_iter = all_indices_.begin() + rank_ * local_chunk_count_;
+    auto end_index_iter = begin_index_iter + local_chunk_count_;
+    std::copy(begin_index_iter, end_index_iter, back_inserter(chunk_indices_));
+    current_index_ = 0;
+  }
+
+ private:
+  std::vector<size_t> all_indices_;
```
Comment (xzhu1900) [inline torch/csrc/api/include/torch/data/datasets/chunk.h:357]: Do we need to keep all_indices_ as a member variable? It seems we can replace it with a local variable in reset().

File: torch/csrc/api/include/torch/data/datasets/chunk.h (hunk 8 above)
Comment (jaliyae) [reply torch/csrc/api/include/torch/data/datasets/chunk.h:390]: If it is a regular sampler, this could be a hit, but since this is only for chunk selection and it needs to be always thread_safe irrespective of distributed or not. ChunkSelector is used by the preloader threads in the chunkdataset.

File: torch/csrc/api/include/torch/data/datasets/chunk.h (hunk 9 above)
Comment (jaliyae) [reply torch/csrc/api/include/torch/data/datasets/chunk.h:357]: it does not, but don't want to create this again and again in every reset.

Comment (jaliyae) [reply torch/csrc/api/include/torch/data/datasets/chunk.h:357]: my expectation is that chunks are bigger so not many indices here.

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 10]
```diff
@@ -240,6 +249,133 @@ class BatchDataBuffer {
...
+      : ChunkSelector(chunk_count, num_replicas, rank) {
+    size_t index_count =
+        num_replicas_ == 1 ? chunk_count_ : local_chunk_count_ * num_replicas_;
+    all_indices_.resize(index_count);
+    std::iota(std::begin(all_indices_), std::end(all_indices_), 0);
+    if (num_replicas_ > 1)
+    {
+      for (size_t i = chunk_count; i < index_count; ++i) {
+        all_indices_[i] =
+            i % chunk_count_; // we are adding some more chunks to make all
```
Comment (xzhu1900) [inline torch/csrc/api/include/torch/data/datasets/chunk.h]: Curious, so we will have duplicated chunk index and load duplicated chunk data ? Will this be a problem in training?

File: torch/csrc/api/include/torch/data/datasets/chunk.h (hunk 8 above)
Comment (xzhu1900) [reply torch/csrc/api/include/torch/data/datasets/chunk.h:390]: D'oh! yeah, you are right.

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 11]
```diff
@@ -242,6 +251,135 @@ class BatchDataBuffer {
...
+    end_index_ = begin_index_ + local_chunk_count_;
+    chunk_index_ = begin_index_;
+  }
+
+  optional<size_t> next() override {
+    size_t idx = chunk_index_.fetch_add(1, std::memory_order_relaxed);
+    if (idx < end_index_) {
+      return idx % chunk_count_;
+    } else {
+      return nullopt; /// Select chunks randomly. The chunk order shuffled at
```
Comment (thiagocrepaldi) [inline torch/csrc/api/include/torch/data/datasets/chunk.h]: NIT: copy&paste comment issue? Chunks are not randomly selected for SequentialChunkSelector

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 12]
```diff
@@ -341,11 +476,15 @@ class ChunkDataset final
     // free workers from previous reset if there is any.
     free_workers();
     preload_threads_.clear();
-
+
```
Comment (thiagocrepaldi) [inline torch/csrc/api/include/torch/data/datasets/chunk.h]: NIT: remove whitespaces/tabs from empty lines. Is clang-format enabled on your vscode?

File: torch/csrc/api/include/torch/data/datasets/chunk.h [hunk 13]
```diff
@@ -242,6 +251,135 @@ class BatchDataBuffer {
...
+    chunk_indices_.reserve(local_chunk_count_);
+    auto begin_index_iter = all_indices_.begin() + rank_ * local_chunk_count_;
+    auto end_index_iter = begin_index_iter + local_chunk_count_;
+    std::copy(begin_index_iter, end_index_iter, back_inserter(chunk_indices_));
+    current_index_ = 0;
+  }
+
+ private:
+  std::vector<size_t> all_indices_;
+  std::vector<size_t> chunk_indices_;
```
Comment (thiagocrepaldi) [inline torch/csrc/api/include/torch/data/datasets/chunk.h]: Depending on the size of `chunk_indices`,  we could discard it to save memory and time during vector copying from `all_indices_` to `chunk_indices_`

We can keep a single copy of indices at `all_indices_` and each replica would keep two iterators for the beginning (`rank_ * local_chunk_count_`) and ending (`rank_ * local_chunk_count_ + local_chunk_count_`) of its own chunks using. Indexing would be `all_indices_[rank_ * local_chunk_count_ + current_index_]`

File: torch/csrc/api/include/torch/data/datasets/chunk.h (hunk 10 above)
Comment (jaliyae) [reply torch/csrc/api/include/torch/data/datasets/chunk.h]: Thank you for the great question. We figured an important solution from this.

File: torch/csrc/api/include/torch/data/datasets/chunk.h (hunk 11 above)
Comment (jaliyae) [reply torch/csrc/api/include/torch/data/datasets/chunk.h]: thanks.

File: torch/csrc/api/include/torch/data/datasets/chunk.h (hunk 13 above)
Comment (jaliyae) [reply torch/csrc/api/include/torch/data/datasets/chunk.h]: Thank you for the suggestion. Fixed it.

File: torch/csrc/api/include/torch/data/datasets/chunk.h (hunk 12 above)
Comment (jaliyae) [reply torch/csrc/api/include/torch/data/datasets/chunk.h]: Thank you. It is enabled.

File: test/cpp/api/dataloader.cpp [hunk 14]
```diff
@@ -97,17 +97,19 @@ TEST(DataTest, ChunkDataSetWithInvalidInitParameter) {
 
   auto initialization_function =
       [&](size_t preloader_count, size_t batch_size, size_t cache_size) {
+        std::shared_ptr<datasets::ChunkSelector> chunk_selector =
```
Comment (xzhu1900) [inline test/cpp/api/dataloader.cpp:100]: One question I had since the first iteration was that, whether it is a good idea to make it a shared pointer. This gives the user access to change chunk_selector while our chunkDataset is still using it. For example, the user may unintentionally call reset while the epoch is not fully exhausted yet. 

Maybe we can consider make a copy of the chunk_selector inside chunkdataset?
